packages/upstash-redis/src/index.ts

```diff
@@ -0,0 +1,212 @@
+    baseKeyPrefix + mergedOptions.verificationTokenKeyPrefix
+
+  const setObjectAsJson = async (key: string, obj: any) =>
+    await client.set(key, JSON.stringify(obj))
+
+  const setAccount = async (id: string, account: AdapterAccount) => {
+    const accountKey = accountKeyPrefix + id
+    await setObjectAsJson(accountKey, account)
+    await client.set(accountByUserIdPrefix + account.userId, accountKey)
+    return account
+  }
+
+  const getAccount = async (id: string) => {
+    const response = await client.get(accountKeyPrefix + id)
+    if (!response.data) return null
+    return reviveFromJson(response.data)
+  }
+
+  const setSession = async (id: string, session: AdapterSession) => {
+    const sessionKey = sessionKeyPrefix + id
+    await setObjectAsJson(sessionKey, session)
+    await client.set(sessionByUserIdKeyPrefix + session.userId, sessionKey)
+    return session
+  }
+
+  const getSession = async (id: string) => {
+    const response = await client.get(sessionKeyPrefix + id)
+    if (!response.data) return null
+    return reviveFromJson(response.data)
+  }
+
+  const setUser = async (id: string, user: AdapterUser) => {
+    await setObjectAsJson(userKeyPrefix + id, user)
+    await client.set(`${emailKeyPrefix}${user.email as string}`, id)
+    return user
+  }
+
+  const getUser = async (id: string) => {
+    const response = await client.get(userKeyPrefix + id)
+    if (!response.data) return null
+    return reviveFromJson(response.data)
+  }
+
+  return {
+    async createUser(user: Omit<AdapterUser, "id">): Promise<AdapterUser> {
+      const id = uuid()
+      // TypeScript thinks the emailVerified field is missing
+      // but all fields are copied directly from user, so it's there
+      // @ts-expect-error
+      return await setUser(id, { ...user, id })
+    },
+    getUser,
+    async getUserByEmail(email: string): Promise<AdapterUser | null> {
+      const emailResponse = await client.get(emailKeyPrefix + email)
+      if (!emailResponse.data) return null
+      return await getUser(emailResponse.data)
+    },
+    async getUserByAccount(
+      account: Pick<AdapterAccount, "provider" | "providerAccountId">
+    ): Promise<AdapterUser | null> {
+      const dbAccount = await getAccount(
+        `${account.provider}:${account.providerAccountId}`
+      )
+      if (!dbAccount) return null
+      return await getUser(dbAccount.userId)
+    },
+    async updateUser(updates: Partial<AdapterUser>): Promise<AdapterUser> {
+      const userId = updates.id as string
+      const user = await getUser(userId)
+      return await setUser(userId, { ...user, ...updates })
+    },
+    async linkAccount(account: AdapterAccount): Promise<AdapterAccount> {
+      const id = `${account.provider}:${account.providerAccountId}`
+      return await setAccount(id, { ...account, id })
+    },
+    async createSession(session: {
+      sessionToken: string
+      userId: string
+      expires: Date
+    }): Promise<AdapterSession> {
```

Comment:
```suggestion
    async createSession(session) {
```

Typings like these on the returned adapter methods should not be necessary, they are inferred from the `Adapter` interface added on line 45. Could you remove them?